packages/kbn-typed-react-router-config/src/create_router.ts

```diff
@@ -110,6 +116,11 @@ export function createRouter<TRoutes extends RouteMap>(routes: TRoutes): Router<
       throw new Error(errorMessage);
     }
 
+    const hasExactMatch = matches.some((match) => match.match.isExact);
+    if (!hasExactMatch) {
+      throw new NotFoundRouteException('No route was matched');
+    }
```

Comment:
I think this code should (as in, arguably) already throw on line 116. Maybe that logic is not sufficient but I'd prefer it to be in one place. Also, there are tests for matchRoutes, also where it (already) throws an exception. We should add tests for not found exceptions. Maybe tests will also make it clear why the current logic is not sufficient.

Reply:
@dgieselaar from what I could see the exception on line 116 does NOT throw because we have two routes defined for the `/` path. One on the [apm_route_config](https://github.com/cauemarcondes/kibana/blob/apm-404/x-pack/plugins/apm/public/components/routing/apm_route_config.tsx#L67) and another on [home/index](https://github.com/cauemarcondes/kibana/blob/apm-404/x-pack/plugins/apm/public/components/routing/home/index.tsx#L100).

Take a look at this picture that shows two matches for the following path: `http://localhost:5601/foo/app/apm/foo`
<img width="763" alt="Screenshot 2023-01-31 at 9 32 34 AM" src="https://user-images.githubusercontent.com/55978943/215789555-99055453-04a3-4392-ad37-56f86855bb82.png">

That's why I added the new NotFoundException.

Reply:
I think they should both throw a NotFoundException. But we can do that in a follow-up. Can you create a ticket for that?